elfo-test/src/util.rs

```diff
@@ -0,0 +1,73 @@
+use elfo_core::{msg, Envelope, Message, Request, ResponseToken};
+
+/// Extracts message with type [`M`] from [`Envelope`], panics otherwise.
+#[track_caller]
+pub fn extract_msg<M: Message>(msg: Envelope) -> M {
+    msg!(match msg {
+        msg @ M => msg,
+        msg => panic!(
+            "unexpected message: expected {}, got: {:?}",
+            elfo_core::dumping::extract_name_by_type::<M>(),
+            msg
+        ),
+    })
+}
+
+/// Extracts request message with type [`R`] from [`Envelope`], panics
+/// otherwise.
+#[track_caller]
+pub fn extract_req<R: Request>(msg: Envelope) -> (R, ResponseToken<R>) {
+    msg!(match msg {
+        (req @ R, token) => (req, token),
+        msg => panic!(
+            "unexpected request: expected {}, got: {:?}",
+            elfo_core::dumping::extract_name_by_type::<R>(),
+            msg
+        ),
+    })
+}
+
+#[cfg(test)]
+mod tests {
+    use elfo_core::{_priv::MessageKind, message, scope::Scope, ActorMeta, Addr};
+
+    use super::*;
+
+    #[message]
+    #[derive(PartialEq)]
+    struct TestMessage;
+
+    #[message(ret = ())]
+    #[derive(PartialEq)]
+    struct TestRequest;
+
+    #[test]
+    fn extract_message() {
+        create_scope().sync_within(|| {
+            let envelop =
+                Envelope::new(TestMessage, MessageKind::Regular { sender: Addr::NULL }).upcast();
+            assert_eq!(extract_msg::<TestMessage>(envelop), TestMessage);
+        });
+    }
+
+    #[test]
+    fn extract_request() {
+        create_scope().sync_within(|| {
+            let envelop =
+                Envelope::new(TestRequest, MessageKind::RequestAll(ResponseToken::test())).upcast();
```

Comment:
It's possible to use here `MessageKind::Regular` to avoid modification of `ResponseToken` API